nltk/translate/ribes_score.py

```diff
@@ -0,0 +1,311 @@
+# -*- coding: utf-8 -*-
+# Natural Language Toolkit: RIBES Score
+#
+# Copyright (C) 2001-2015 NLTK Project
+# Contributors: Katsuhito Sudoh, Liling Tan, Kasramvd, J.F.Sebastian, Mark Byers, ekhumoro, P. Ortiz
+# URL: <http://nltk.org/>
+# For license information, see LICENSE.TXT
+""" RIBES score implementation """
+
+from itertools import islice
+import math
+
+from nltk import ngrams
+
+
+def ribes(references, hypothesis, alpha=0.25, beta=0.10):
+    """
+    The RIBES (Rank-based Intuitive Bilingual Evaluation Score) from 
+    Hideki Isozaki, Tsutomu Hirao, Kevin Duh, Katsuhito Sudoh and 
+    Hajime Tsukada. 2010. "Automatic Evaluation of Translation Quality for 
+    Distant Language Pairs". In Proceedings of EMNLP. 
+    http://www.aclweb.org/anthology/D/D10/D10-1092.pdf 
+    
+    The generic RIBES scores used in shared task, e.g. Workshop for 
+    Asian Translation (WAT) uses the following RIBES calculations:
+    
+        RIBES = kendall_tau * (alpha**p1) * (beta**bp)
+    
+    Please note that this re-implementation differs from the official
+    RIBES implementation and though it emulates the results as describe
+    in the original paper, there are further optimization implemented 
+    in the official RIBES script.
+    
+    Users are encouraged to use the official RIBES script instead of this 
+    implementation when evaluating your machine translation system. Refer
+    to http://www.kecl.ntt.co.jp/icl/lirg/ribes/ for the official script.
+    
+    :param reference: a reference sentence
+    :type reference: list(str)
+    :param hypothesis: a hypothesis sentence
+    :type hypothesis: list(str)
+    :param alpha: hyperparameter used as a prior for the unigram precision.
+    :type alpha: float
+    :param beta: hyperparameter used as a prior for the brevity penalty.
+    :type beta: float
+    :param character_based: a flag to emulate the original character-based RIBES
+    :type character_based: boolean 
+    """
+    _best_ribes = -1.0
+    # Calculates RIBES for each reference and returns the best score.
+    for reference in references:
+        # Collects the *worder* from the ranked correlation alignments.
+        worder = word_rank_alignment(reference, hypothesis)
+        nkt = kendall_tau(worder)
+            
+        # Calculates the brevity penalty
+        bp = min(1.0, math.exp(1.0 - 1.0 * len(reference)/len(hypothesis)))
```

Comment:
`min(...)` is right here because brevity penalty penalizes short hypothesis because `precision = 1.0 * n / len(hyp)` so the shorter the `len(hyp)` the higher the precision. This example will illustrate the penalty:

```
>>> import math
>>> lenref, lenhyp = 10, 6
>>> n = 5
>>> precision = 1.0 * n / lenhyp
>>> precision
0.8333333333333334
>>> bp = min(1.0, math.exp(1.0 - 1.0 * lenref/lenhyp))
>>> bp
0.513417119032592
```



And if the hypothesis is short, precision is high by it didn't match the longer reference so it needs to be penalize

```
>>> lenref, lenhyp = 10, 2
>>> n = 1
>>> precision = 1.0 * n / lenhyp
>>> precision
0.5
>>> bp = min(1.0, math.exp(1.0 - 1.0 * lenref/lenhyp))
>>> bp
0.01831563888873418
```